src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/CSharpDiagnosticWorkerBase.cs

```diff
@@ -0,0 +1,82 @@
+            FileSystemHelper fileSystemHelper)
+    {
+        _workspace = workspace;
+        _fileSystemHelper = fileSystemHelper;
+        _fileSystemMatcher = _fileSystemHelper.BuildExcludeMatcher();
+    }
+
+    public abstract bool AnalyzersEnabled { get; }
+
+    public abstract Task<IEnumerable<Diagnostic>> AnalyzeDocumentAsync(Document document, CancellationToken cancellationToken);
+
+    public abstract Task<IEnumerable<Diagnostic>> AnalyzeProjectsAsync(Project project, CancellationToken cancellationToken);
+
+    public abstract Task<ImmutableArray<DocumentDiagnostics>> GetAllDiagnosticsAsync();
+
+    public abstract Task<ImmutableArray<DocumentDiagnostics>> GetDiagnostics(ImmutableArray<string> documentPaths);
+
+    public abstract ImmutableArray<DocumentId> QueueDocumentsForDiagnostics(IEnumerable<Document> documents, AnalyzerWorkType workType);
+
+    public ImmutableArray<DocumentId> QueueDocumentsForDiagnostics()
+    {
+        IEnumerable<Document> documents = _workspace.CurrentSolution.Projects.SelectMany(x => x.Documents);
+        documents = FilterDocuments(documents);
+        return QueueDocumentsForDiagnostics(documents, AnalyzerWorkType.Background);
+    }
+
+    public ImmutableArray<DocumentId> QueueDocumentsForDiagnostics(ImmutableArray<ProjectId> projectId)
+    {
+        IEnumerable<Document> documents = projectId.SelectMany(projectId => _workspace.CurrentSolution.GetProject(projectId).Documents);
+        documents = FilterDocuments(documents);
+        return QueueDocumentsForDiagnostics(documents, AnalyzerWorkType.Background);
+    }
+
+    protected ImmutableArray<DocumentId> QueueDocumentsForDiagnostics(Project project)
+    {
+        IEnumerable<Document> documents = project.Documents;
+        documents = FilterDocuments(documents);
+        return QueueDocumentsForDiagnostics(documents, AnalyzerWorkType.Background);
+    }
+
+    protected ImmutableArray<DocumentId> QueueDocumentsForDiagnostics(IEnumerable<DocumentId> documentsIds, AnalyzerWorkType workType)
+    {
+        IEnumerable<Document> documents = documentsIds.Select(x => _workspace.CurrentSolution.GetDocument(x));
+        documents = FilterDocuments(documents);
+        return QueueDocumentsForDiagnostics(documents, workType);
+    }
+
+    private ImmutableArray<Document> FilterDocuments(IEnumerable<Document> documents) =>
+        documents
+            .Where(x => !FileSystemHelper.AutoGenerated(x.FilePath))
+            .Select(x =>
+            {
+                using var sw = new StreamWriter("logs.txt");
+                var path = FileSystemHelper.GetRelativePath(x.FilePath, _fileSystemHelper.TargetDirectory);
+                var match = _fileSystemMatcher.Match(path).HasMatches;
+                sw.WriteLine($"{path}: {match}");
+                return x;
+            })
+            .Where(x => !_fileSystemMatcher.Match(FileSystemHelper.GetRelativePath(x.FilePath, _fileSystemHelper.TargetDirectory)).HasMatches)
```

Comment:
When files are added to the test workspace they aren't properly rooted under the TargetDirectory. Easiest fix for the tests would be to assume the FilePath is relative if null is returned from GetRelativePath.

```suggestion
            .Where(x => !_fileSystemMatcher.Match(FileSystemHelper.GetRelativePath(x.FilePath, _fileSystemHelper.TargetDirectory) ?? x.FilePath).HasMatches)
```